feat: Inner Computer + DAO contracts, `getTXOs` query API, DB schema/query improvements, and CI/test reliability hardening (previously Inner Computer and DAO)

ltardivo opened the pull request:
### Summary

This PR delivers major new capabilities to the **Inner Computer** and introduces the first **DAO (Decentralized Autonomous Organization) contract** implementation. It includes a more powerful and flexible querying layer (`getTXOs`), important database schema and indexing improvements for block-aware queries, refined `mod` propagation semantics, and a set of supporting changes that significantly improve test reliability and developer experience.

The changes enable richer on-chain object interaction patterns (revision tracing, ancestry, balance queries, etc.) while making complex DAO-style governance contracts practical.

### Key Changes

**New DAO Contracts & Inner Computer Features**
- Added the new `@bitcoin-computer/dao-contracts` workspace package.
- Smart contracts now have built-in, read-only access to revision graphs, transaction outputs, ancestry tracing, and balance queries (via the Inner Computer) without modifying on-chain state.
- Updated DAO contract implementation (e.g., `validVotes`, amount summation in `Vote` class, removal of `intersect`, cleaner `validate` logic).

**Query Layer & `getTXOs`**
- Introduced `getTXOs(...)` — a more flexible, polymorphic replacement/augmentation for `computer.query(...)`.
- Supports rich filtering by `rev`, `publicKey`, `mod`, `blockHash`, `blockHeight`, `blockIndex`, `isObject`, `isSpent`/`isConfirmed`, etc.
- Improved documentation and added comprehensive tests for the new query capabilities.
- Renamed `spent` → `isSpent` and added `isConfirmed` query option for clarity.

**Database Schema & Performance**
- Added `blockHeight` and `blockIndex` columns to the `Output` table.
- Updated `packages/node/scripts/create-indexes.ts` with the corresponding indexes:
  - `idx_output_blockheight`
  - `idx_output_blockindex`
- These enable efficient block-range and ordering queries that power the new `getTXOs` functionality and future use cases.

**Mod Propagation Semantics (breaking)**
- `mod` is now propagated to **all descendants** of an object creation (previously only set on direct `computer.new`/`encode` calls). This improves consistency for complex object graphs.

**Testing & CI Infrastructure Improvements**
- Migrated `scripts/check-obfuscation.sh` → `scripts/check-obfuscation.js` (Node.js version with better error handling, staged-file git integration, and monorepo-aware dist checking for proper minification/obfuscation).
- Major improvements to `scripts/rerun-failed.js`:
  - Proper `.env` loading (with Docker vs host environment adjustments for `POSTGRES_HOST`, `BITCOIN_RPC_HOST`, etc.).
  - Robust handling of empty/invalid `test-results.json` files.
  - Better path resolution and test title cleaning for Mocha `--grep`.
  - Cleaner execution when run from package directories or monorepo root.
- Upgraded `mocha` to `^11.0.0` (and `@types/mocha` where applicable) in `nodejs-template` and `swap` packages.
- Updated `packages/node/scripts/fund.ts` to use the newer `computer.rpc(...)` API (instead of `rpcCall`).

**Dependencies & Tooling**
- Large `package-lock.json` update reflecting the new `dao-contracts` workspace package, updates across the Parcel ecosystem, addition of supporting packages (Rollup plugins, Puppeteer browsers, React DnD, Remix Router, etc.), and various transitive dependency refreshes. This supports both the new contracts and ongoing work in explorer/components.

**Documentation**
- Refined docs for `getUTXOs` / `getOTXOs` / `getOUTXOs` (now presented as convenient wrappers over `getTXOs`).
- Updated Postgres/DB documentation and various README improvements.

### Breaking Changes / Migration Notes

1. **Database schema change** — New columns + indexes. Run the updated `create-indexes.ts` (or equivalent migration) on existing databases.
2. **`mod` propagation** now applies to the full descendant chain.
3. **`computer.query`** usage should migrate toward `getTXOs` for new filtering needs (especially anything block- or confirmation-related).
4. Internal RPC helper renamed in scripts (`rpcCall` → `rpc`).

### Motivation

These changes make sophisticated DAO governance and cross-contract state synchronization practical on Bitcoin/Litecoin. The new query capabilities and block indexing remove previous limitations when building complex, stateful on-chain applications. The CI/test improvements (especially the more robust failure rerun script and obfuscation checker) were necessary to maintain velocity and reliability while landing this large set of changes.

### Testing

- Added and updated tests for `getTXOs` (by rev, publicKey, mod, block fields, etc.).
- DAO contract tests validate core voting/proposal flows.
- CI improvements were validated by exercising the new `rerun-failed.js` and obfuscation check paths.
- Full test suite run with the updated Mocha version.

ltardivo committed:
Adds DAO project

ltardivo committed:
Adds more errors in the troubleshooting section

ltardivo committed:
Upload bundles

ltardivo committed:
Renaming

ltardivo committed:
Modifies error docs description

ltardivo committed:
Updates dao contract. Extracts validVotes function. Stores amount sum in Vote class

ltardivo committed:
Runs lint:fix on build

ltardivo committed:
Updates module specifiers by descendant relationship

ltardivo committed:
Set mod only when previous
is undefined

ltardivo committed:
Merge branch 'staging' into feat/update-module-specifiers

ltardivo committed:
Adds BCN_QUERY_LIMIT to .env.example

ltardivo committed:
Adds getTxos function

ltardivo committed:
Refactor types. Changes Output table fields

ltardivo committed:
Enables querying by rev. Improves docs for getTxos

ltardivo committed:
Adds test for getTxos by rev, publicKey and isObject

ltardivo committed:
Fixes test on mainnet network

ltardivo committed:
Renaming

ltardivo committed:
Adds more test to get-txos.test.ts

ltardivo committed:
Renames expHash to exp

ltardivo committed:
Renames spent to isSpent. Adds query option isConfirmed

ltardivo committed:
Changes mod semantics in function calls, propagates the mod from the mod of the object creation

ltardivo committed:
Makes getTXOs polymorphic

ltardivo committed:
Improves docs. Adds missing node endpoints. Refers to docs from README.md

ltardivo committed:
Adds blockHeight and blockIndex to Outputs table

ltardivo committed:
Fixes postgres docker image to version 15

ltardivo committed:
Adds blockHeight and blockIndex to DbOutput type

ltardivo committed:
Adds test for getTXOs with blockHash and blockIndex

ltardivo committed:
Merge feat/dao to feat/update-module-specifiers

ltardivo committed:
Upload bundles

ltardivo committed:
Refactors DAO. Removes intersect, query revision in validate function

ltardivo committed:
Adds README.md to DAO project

ltardivo committed:
Removes duplicated prototype in types

ltardivo committed:
Exports computer to globals

ltardivo committed:
Adds keywords to lib package.json

ltardivo committed:
Adds tokenRoot to Vote class to enforce token lineage

ltardivo committed:
Refactor dao-contract. Returns accepted and rejected number of votes

ltardivo committed:
Updates README.md

ltardivo committed:
Renames rpcCall to rpc

ltardivo committed:
Revert change. Imports contract in token class

ltardivo committed:
Renames rpcCall into rpc

ltardivo committed:
Adds getAncestors to Computer API

ltardivo committed:
Removes sleep. Adds validRevVotes function for testing

ltardivo committed:
More work on development docs

ltardivo committed:
Renames type from _Balance to Balance

ltardivo committed:
Adds section to docs on querying functions available in a Smart Contract

ltardivo committed:
Type fixes

ltardivo committed:
Adds test for voting

ltardivo committed:
Upload bundles

ltardivo committed:
More work on docs

ltardivo committed:
Upload bundles

ltardivo committed:
Merge staging into branch

ltardivo committed:
Resolve merge conflicts

ltardivo committed:
Validates staging files in pre-commit

ltardivo committed:
Adds checks for staged files.  to nodeJs

ltardivo committed:
Do not restore staged files on bundles checks

ltardivo committed:
Fixes version mismatch between ESLint v9 and the TypeScript parser plugin

ltardivo committed:
Replaces computer.query by computer.getOUTXOs

ltardivo committed:
Removes utxos endpoint from docs. Adds get-txos docs

ltardivo committed:
Fixes Parcel version to 2.15.4

ltardivo committed:
Removes package glob in rerun-failed.js

ltardivo committed:
Uploads package-lock.json

ltardivo committed:
Replaces getFormattedUtxos function by getUTXOs

ltardivo committed:
Builds source before testing

ltardivo committed:
Adds .mocharc.json for rerun failed tests

ltardivo committed:
Fixes rerun script

ltardivo committed:
Revert eslint.config.mjs changes

ltardivo committed:
Load env variables from .env

ltardivo committed:
Upgrades mocha to version 11.0.0

ltardivo committed:
Uploads bundles

ltardivo committed:
Extracts components

ltardivo committed:
Fixes deploy script invalid path

ltardivo committed:
Enables next lint

ltardivo committed:
Uploads package-lock.json

ltardivo committed:
Removes console

ltardivo committed:
Upload bundles

ClemensLey commented on packages/chess-contracts/src/chess-contract.ts:
```diff
@@ -211,10 +211,12 @@ export class ChessContractHelper {
     let paid = 0n
     while (paid < Number(this.satoshis) / 2 && utxos.length > 0) {
-      const { txId, vout, satoshis } = utxos.pop()!
+      const { rev, satoshis } = utxos.pop()!
```
We should consider using parseRev here. If it is not exported we should export it

ClemensLey committed:
Fix formatting in README.md for clarity

ClemensLey committed:
Refactor GetTXOs types and improve documentation of getTXOs function

ClemensLey committed:
Update Postgres database documentation

Reorganize Postgres database section and add credentials information.

ClemensLey committed:
Update visibility in development.md

Remove visibility setting from development documentation

ClemensLey commented on packages/lib/test/lib/computer/get-network.test.ts:
```diff
@@ -4,7 +4,7 @@ import { expect } from '../../utils/index.js'
   it('Should return the network', async () => {
     // Create wallet with a specific network
-    const computer = new Computer({ network: 'mainnet' })
+    const computer = new Computer({ network: 'mainnet', mode: 'prod' })
```
We should consider making this `dev`

ClemensLey commented:
In case we don't have documentation for `getUTXOs`, `getOTXOs` and `getOUTXOs` we should add them. They all should have their own page/file but should just contain their definition in `computer` (to point out that they are syntactic sugar) and a reference to the documentation of `getTXOs`.

ClemensLey commented on packages/dao-contracts/src/dao-contract.ts:
```diff
@@ -0,0 +1,95 @@
+      const ancestorsSet = new Set<string>(ancestors)
+      ancestorsSet.delete(voteTxId)
+
+      if ([...ancestorsSet].some(voteTxIdsSet.has, voteTxIdsSet)) {
```
I think this line would be easier to read as 

```
if ([...ancestorsSet].some((element) => voteTxIdsSet.has(element))) { ... }
```

ClemensLey commented on packages/dao-contracts/src/dao-contract.ts:
```diff
@@ -0,0 +1,95 @@
+      tokenRoot: tokens[0]._root,
+    })
+    // check that all tokens have the same root
+    const tokenRoots = new Set(tokens.map((t) => t._root))
```
In terms of readability I think it would be nicer if this check were at the top of the constructor.

ClemensLey commented on packages/dao-contracts/src/dao-contract.ts:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  private async validVotes(): Promise<Vote[]> {
+    const proposalVotes = await this.proposalVotes()
```
Unless it is important for testing, I would suggest to inline `proposalVotes()` here because it is not so clear from the name what the function does. Otherwise we should try to find a better name.

ClemensLey commented on packages/dao-contracts/src/dao-contract.ts:
```diff
@@ -0,0 +1,95 @@
+    )
+  }
+
+  async validRevVotes(): Promise<string[]> {
```
As far as I can see this function can be removed

ClemensLey commented on packages/dao-contracts/src/dao-contract.ts:
```diff
@@ -0,0 +1,95 @@
+
+    return [...resolved].filter(
+      (r: Vote) => r.electionId === this._id && r.tokenRoot === this.tokenRoot,
+    )
```
In addition we have to check that all transactions with id in `proposalVote` have the correct expression (for example using `computer.decode` with the inner computer). Otherwise someone can just use a class that sets `electionId` and `tokenRoot` as desired and `tokenAmount` to an arbitrarily large number.

ClemensLey replied:
My suggestion would be to come back to this comment once we have refactored according to the other comments.

ltardivo committed:
Upload bundles

ltardivo committed:
Merge resolution

ClemensLey committed:
Refine `getOTXOs` function documentation

Updated the description of the `getOTXOs` function for clarity and conciseness.

ClemensLey committed:
Rename getUOTXOs to getOUTXOs in docs

Updated function name and description for clarity.

ClemensLey committed:
Fix formatting of return description in getTxos.md

ClemensLey committed:
Revise `getUTXOs` function documentation

ClemensLey reviewed (approved):
Looks good to me

ltardivo committed:
Fixes types
